examples_tests/43.AllocatorTest/main.cpp

```diff
@@ -1,53 +1,219 @@
 	{
-		void* resSpc = _IRR_ALIGNED_MALLOC(core::GeneralpurposeAddressAllocator<uint32_t>::reserved_size(alignof(uint32_t), 16u, 1u), _IRR_SIMD_ALIGNMENT);
-		assert(resSpc != nullptr);
-		auto alctr = core::GeneralpurposeAddressAllocator<uint32_t>(resSpc, 0u, 0u, alignof(uint32_t), 16u, 1u);
-
-		for (int i = 0; i < 16u; i++)
-		{
-			auto addr = alctr.alloc_addr(1u, 1u);
-			if (addr == alctr.invalid_address)
-				os::Printer::print("invalid adress");
-			else
-				os::Printer::print(std::to_string(addr));
+	}
+	
+	//struct for holding temporary data (inputs of `multi_alloc_addr` i `multi_free_addr`)
+	struct EntryForFrameData
+	{
+		uint32_t outAddr[maxAllocs] = { ~0u };
+		uint32_t sizes[maxAllocs] = { 0u };
+		uint32_t alignments[maxAllocs] = { 0u };
+	} perFrameData;
+	
+	void executeAllocatorTest()
+	{
+		os::Printer::log("Executing Pool Allocator Test!", ELL_INFORMATION);
+	
+		for (size_t i = 0; i < creationParameters.multiAllocCnt; ++i)
+			executeForFrame();
+	}
+
+private:
+	// function to test allocator
+		// pick random alignment, rand buffer size (between 2kb and 2GB), random offset (less than buffer size), random alignment offset (less than alignment) and other parameters randomly
+
+		// allocate reserved space (for allocator state)
+
+		// create address allocator
+
+		// randomly decide the number of iterations of allocation and reset
+			// declare `core::vector` to hold allocated addresses and sizes
+
+			// randomly decide how many `multi_allocs` to do
+				// randomly decide how many allocs in a `multi_alloc` NOTE: must pick number less than `traits::max_multi_alloc`
+					// randomly decide sizes (but always less than `address_allocator_traits::max_size`)
+				// record all successful alloc addresses to the `core::vector`
+
+				// run random dealloc function
+			//
+
+			// run random dealloc function
+
+			// randomly choose between reset and freeing all `core::vector` elements
+				// reset
+			// ELSE
+				// free everything with a series of multi_free
+	void executeForFrame()
+	{
+		uint32_t randMaxAlign = rng.getRndMaxAlign();
+		uint32_t randAddressSpaceSize = rng.getRndBuffSize();
+		uint32_t randAlignOffset = rng.getRandomNumber(0u, randMaxAlign - 1u);
+		uint32_t randOffset = rng.getRandomNumber(0u, randAddressSpaceSize - 1u); //?
+		uint32_t randBlockSz = rng.getRandomNumber(0u, randAddressSpaceSize - 1u); //?
+
+		uint32_t allocationAmountForSingeFrameTest = rng.getRndAllocCnt();
+		
+		const auto reservedSize = PoolAddressAllocator::reserved_size(randMaxAlign, randAddressSpaceSize, randBlockSz); // 3rd parameter onward is custom for each address alloc type
+		void* reservedSpace = _IRR_ALIGNED_MALLOC(reservedSize, _IRR_SIMD_ALIGNMENT);
+		auto poolAlctr = PoolAddressAllocator(reservedSpace, randOffset, randAlignOffset, randMaxAlign, randAddressSpaceSize, randBlockSz);
+
+		// randomly decide how many `multi_allocs` to do
+		const uint32_t multiAllocCnt = rng.getRandomNumber(1u, 5u); //range?
```

Comment:
as many as reasonable time-wise for the test, probably in the thousands or millions though.